Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,248 @@
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        timeoutInterval: TimeInterval = 1.0,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard case .remote(_, let urlString) = package else {
+            return completion(.failure(RegistryError.invalidPackage))
+        }
+
+        guard let url = URL(string: "https://\(urlString)") else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = URLRequest(url: url, cachePolicy: .returnCacheDataElseLoad, timeoutInterval: timeoutInterval)
+        request.httpMethod = "HEAD"
+
+        let session = URLSession(configuration: configuration)
+
+        let task = session.dataTask(with: request) { _, response, error in
+            if let error = error {
+                completion(.failure(error))
+            } else if let response = response as? HTTPURLResponse,
+                      response.statusCode == 303,
+                      response.allHeaderFields["Content-Version"] as? String == "1",
+                      let location = response.allHeaderFields["Location"] as? String,
+                      let redirectedURL = URL(string: location)
+            {
+                let manager = RegistryManager(registryBaseURL: redirectedURL)
+                completion(.success(manager))
+            } else {
+                completion(.failure(RegistryError.invalidResponse))
+            }
+        }
+        task.resume()
+    }
+
+    public func fetchVersions(of package: PackageReference, completion: @escaping (Result<[Version], Error>) -> Void) {
+        var components = URLComponents(url: registryBaseURL, resolvingAgainstBaseURL: true)!
+        components.path = "/\(package.identity)"
+
+        guard let url = components.url else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.addValue("application/vnd.swift.registry.v1+json", forHTTPHeaderField: "Accept")
+
+        let task = self.session.dataTask(with: request) { data, response, error in
+            if let error = error {
+                completion(.failure(error))
+            } else if let response = response as? HTTPURLResponse,
+                      response.statusCode == 200,
+                      response.allHeaderFields["Content-Version"] as? String == "1",
+                      response.allHeaderFields["Content-Type"] as? String == "application/json",
+                      let data = data,
+                      case .dictionary(let payload) = try? JSON(data: data),
+                      case .dictionary(let releases) = payload["releases"]
+            {
+                let versions = releases.keys.compactMap { Version(string: $0) }.sorted(by: >)
```

Comment:
Should we filter out releases that are problematic? (i.e., `problem` is present)

Reply:
My first thought was to punt on this because the semantics / logistics for unpublishing aren't entirely settled, but I think you're right that we can, as a first step, filter out anything with a `problem` field. I've done that with aab50bd3b20017c6e568190b73432c5ba97eda84.

For posterity, the relevant discussion is in [4.1. List package releases](https://github.com/apple/swift-package-manager/blob/872ce5c7dfcf4df7cb8b636db4aa9707ba6e416e/Documentation/Registry.md#41-list-package-releases).